examples/cfq/utils.py

```diff
@@ -0,0 +1,234 @@
+        datasets: datasets from which the vocabulary should be extracted
+        special_tokens: special tokens to be added in the vocabulary; they are
+                        added at the begginig with PAD at index 0
+        preprocessing_fun: function for applying preprocessing on a dataset
+                           entry; by default the identity function
+                           (no preprocessing)
+    Returns:
+        An ordered dictionary that maps tokens to their IDs in the vocabulary.
+    """
+
+    vocab_dir = 'temp'
+    vocab_path = vocab_dir+'/'+file_name
+    if not os.path.exists(vocab_dir):
+        os.makedirs(vocab_dir)
+    # Try to read the vocabulary
+    if os.path.isfile(vocab_path):
+        with open(vocab_path, 'rb') as vocab_file:
+            vocab = pickle.load(vocab_file)
+            vocab_file.close()
+            return vocab
+
+    tokens_from_datasets = _get_tokens(input_features,
+                                       tokenizer, 
+                                       datasets,
+                                       preprocessing_fun)
+
+    # Count all the tokens.
+    counter = collections.Counter()
+    for tokens in tokens_from_datasets:
+        counter.update(tokens)
+
+    # Add special tokens to the start of vocab.
+    vocab = collections.OrderedDict()
+    for token in special_tokens:
+        vocab[token] = len(vocab)
+
+    # Add all other tokens to the vocab 
+    for token, freq in sorted(
+            # Sort by frequency (from high to low), and then by token string.
+            # This makes sure high frequency tokens get a low token ID.
+            counter.items(),
+            key=lambda token_freq: (-token_freq[1], token_freq[0])):
+        vocab[token] = len(vocab)
+
+    logging.info('Number of unfiltered tokens: %d', len(counter))
+    logging.info('Vocabulary size: %d', len(vocab))
+
+    # Save the vocabulary to disk
+    with open(vocab_path, 'wb') as vocab_file:
+        pickle.dump(vocab, vocab_file)
+        vocab_file.close()
+
+    return vocab
+
+
+def build_tf_hashtable(vocabulary: Dict[bytes, int],
+                       unk_idx: int) -> tf.lookup.StaticHashTable:
+    """Returns a TF lookup table from a vocabulary."""
+    return tf.lookup.StaticHashTable(tf.lookup.KeyValueTensorInitializer(
+        list(vocabulary.keys()), list(vocabulary.values())),
+                                     default_value=unk_idx)
+
+
+def cardinality(dataset: tf.data.Dataset) -> int:
+    """Returns the number of examples in the dataset by iterating over it once."""
+    return dataset.reduce(np.int64(0), lambda x, _: x + 1).numpy()
+
+
+def get_max_length(dataset: tf.data.Dataset, len_fn: Any):
+    """Returns the max length in a dataset
+  Args:
+    dataset: the tensorflow dataset
+    len_fn: function that gets as argument a dataset example and returns the length of that example
+  """
+    # why is the length field of type int32??
```

Comment:
int32 seems like a reasonable datatype for an integer number to me, don't you agree?

Reply:
I just don't know why some fields are int32 and some int64... when I looked at the types there didn't seem to be a consistency...